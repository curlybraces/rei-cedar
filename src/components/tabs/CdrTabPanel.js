export { default as CdrTabPanel } from './CdrTabPanel.jsx';