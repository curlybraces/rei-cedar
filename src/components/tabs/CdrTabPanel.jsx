import modifier from 'mixinsdir/modifier';
import style from './styles/CdrTabPanel.scss';

export default {
  name: 'CdrTabPanel',
  mixins: [modifier],
  props: {
    /**
     * Required string value that shows up on tab header
     */
    name: {
      type: String,
      required: true,
    },
    /**
     * Optional extra reference value to be set on the tab, otherwise
     * the tab name will be used for reference.
     */
    id: String,
  },
  data() {
    return {
      active: false,
      offsetX: 0,
      tabId: this.id || this.name,
      animationDirection: 'default',
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-tab-panel';
    },
    animationHooks() {
      return {
        beforeEnter: this.setEnterStart,
        afterEnter: this.setEnterEnd,
        beforeLeave: this.setLeaveStart,
        afterLeave: this.setLeaveEnd,
        ...this.$listeners,
      };
    },
  },
  methods: {
    setActive(state) {
      this.active = state;
      this.$emit('tab-change', state, this.tabId);
      this.$emit('tabChange', state, this.tabId);
    },
    setAnimationDirection(direction) {
      this.animationDirection = direction;
    },
    setOffsetX(x) {
      this.offsetX = x;
    },
    setEnterStart(element) {
      const el = element;
      el.style.animationDirection = 'reverse';
      el.style.animationTimingFunction = 'cubic-bezier(0.4, 0, 0.68, .06)';
      el.classList.add(this.style[this.animationDirection]);
    },
    setEnterEnd(element) {
      const el = element;
      el.style.animationDirection = '';
      el.classList.remove(this.style[this.animationDirection]);
    },
    setLeaveStart(element) {
      const el = element;
      el.classList.add(this.style[this.animationDirection]);
      el.style.animationTimingFunction = 'cubic-bezier(0.32, 0.94, 0.6, 1)';
    },
    setLeaveEnd(element) {
      const el = element;
      el.classList.remove(this.style[this.animationDirection]);
    },
    handleUpArrowNav() {
      this.$parent.setFocusToActiveTabHeader();
    },
  },
  render() {
    return (
      <transition
        name="fly"
        {...{ on: this.animationHooks }}
      >
        <div
          v-show={this.active}
          aria-hidden={!this.active}
          class={this.modifierClass}
          ref="cdrTabPanelContainer"
          tabindex="0"
          role="tabpanel"
          vOn:keydown_up_prevent={this.handleUpArrowNav}
          key={this.name}
        >
          {this.$slots.default}
        </div>
      </transition>
    );
  },
};
