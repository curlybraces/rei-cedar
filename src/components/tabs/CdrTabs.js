export { default as CdrTabs } from './CdrTabs.jsx';