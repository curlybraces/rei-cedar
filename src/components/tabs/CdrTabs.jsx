import debounce from 'lodash-es/debounce';
import clsx from 'clsx';
import { CdrSpaceOneX } from '@rei/cdr-tokens';
import modifier from '../../mixins/modifier';
import CdrButton from '../button/CdrButton';
import CdrIcon from '../icon/CdrIcon';
import size from '../../mixins/size';
import style from './styles/CdrTabs.scss';

export default {
  name: 'CdrTabs',
  components: {
    CdrButton,
    CdrIcon,
  },
  mixins: [modifier, size],
  props: {
    height: {
      type: String,
      default: '240px',
    },
    activeTab: {
      type: Number,
      default: 0,
    },
  },
  data() {
    return {
      tabs: [],
      underlineOffsetX: 0,
      underlineWidth: 0,
      underlineScrollX: 0,
      overflowScrollX: 0,
      activeTabIndex: 0,
      pages: [],
      pageIndex: 0,
      headerWidth: 0,
      headerScrollWidth: 0,
      tabPages: {},
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-tabs';
    },
    underlineStyle() {
      return {
        transform: `translateX(${this.underlineOffsetX}px)`,
        width: `${this.underlineWidth}px`,
      };
    },
    headerOverflow() {
      return this.headerScrollWidth > this.headerWidth;
    },
    leftPosition() {
      return this.pageIndex ? `-${this.pages[this.pageIndex].offsetLeft}` : 0;
    },
    overflowLeft() {
      return this.pages.length > 0 && this.pageIndex !== 0;
    },
    overflowRight() {
      return this.pages.length > 0 && this.pageIndex !== this.pages.length - 1;
    },
  },
  mounted() {
    this.tabs = (this.$slots.default || [])
      .map(vnode => vnode.componentInstance)
      .filter(tab => tab); // get vue component children in the slot

    this.activeTabIndex = this.getNextTab(this.activeTab);

    if (this.tabs[this.activeTabIndex] && this.tabs[this.activeTabIndex].setActive) {
      this.tabs[this.activeTabIndex].setActive(true);
    }

    this.$nextTick(() => {
      this.calculateOverflow();
      this.updateUnderline();
    });
    // Check for header overflow on window resize for gradient behavior.
    window.addEventListener('resize', debounce(() => {
      this.calculateOverflow();
      this.updateUnderline();
    }, 500));
  },
  methods: {
    getNextTab(startIndex) {
      for (let i = startIndex; i < this.tabs.length; i += 1) {
        if (!this.tabs[i].disabled) {
          return i;
        }
      }

      if (startIndex !== 0) {
        for (let k = 0; k < startIndex; k += 1) {
          if (!this.tabs[k].disabled) {
            return k;
          }
        }
      }

      return -1;
    },
    getPreviousTab(startIndex) {
      for (let i = startIndex; i > -1; i -= 1) {
        if (!this.tabs[i].disabled) {
          return i;
        }
      }

      if (startIndex !== this.tabs.length - 1) {
        for (let k = this.tabs.length - 1; k > startIndex; k -= 1) {
          if (!this.tabs[k].disabled) {
            return k;
          }
        }
      }

      return -1;
    },
    handleClick: debounce(function handleClickCallback(tabClicked) {
      const newIndex = this.tabs.findIndex(tab => tabClicked.name === tab.name);
      this.changeTab(newIndex);
    }, 500, { leading: true, trailing: false }),
    changeTab(newIndex) {
      const oldIndex = this.activeTabIndex;

      if (newIndex > oldIndex) {
        this.tabs[oldIndex].setAnimationDirection('exit-left');
        this.tabs[oldIndex].setActive(false);
        setTimeout(() => {
          this.tabs[newIndex].setActive(true);
          this.tabs[newIndex].setAnimationDirection('enter-right');
        }, 200);
      } else {
        this.tabs[oldIndex].setAnimationDirection('exit-right');
        this.tabs[oldIndex].setActive(false);
        setTimeout(() => {
          this.tabs[newIndex].setActive(true);
          this.tabs[newIndex].setAnimationDirection('enter-left');
        }, 200);
      }
      this.activeTabIndex = newIndex;
      this.updateUnderline();
      this.$refs.cdrTabsHeader.children[this.activeTabIndex].children[0].focus();
    },
    rightArrowNav: debounce(function handleRightArrow() {
      const nextTab = this.getNextTab(this.activeTabIndex + 1);
      const nextTabPage = this.tabPages[nextTab];

      if (nextTabPage !== this.pageIndex) {
        this.underlineWidth = 0;
        this.pageIndex = nextTabPage;
      }
      this.changeTab(nextTab);
    }, 300, { leading: true, trailing: false }),
    leftArrowNav: debounce(function handleLeftArrow() {
      const previousTab = this.getPreviousTab(this.activeTabIndex - 1);
      const previousTabPage = this.tabPages[previousTab];

      if (previousTabPage !== this.pageIndex) {
        this.underlineWidth = 0;
        this.pageIndex = previousTabPage;
      }
      this.changeTab(previousTab);
    }, 300, { leading: true, trailing: false }),
    slideRight() {
      this.pageIndex += 1;
    },
    slideLeft() {
      this.pageIndex -= 1;
    },
    calculateOverflow() {
      if (this.$refs.cdrTabsHeader) {
        this.headerWidth = this.$refs.cdrTabsHeader.offsetWidth;
        this.headerScrollWidth = this.$refs.cdrTabsHeader.scrollWidth;
      }

      if (this.headerOverflow) this.calculatePagination();
      this.updateUnderline();
    },
    calculatePagination() {
      let width = 0; // determines when new page occurs
      let totalWidth = 0; // total width of elements calculated
      const headerElements = Array.from(this.$refs.cdrTabsHeader.children); // tabs
      const endPage = this.headerScrollWidth - this.headerWidth; // end scroll position is known
      const tabSpace = Number(CdrSpaceOneX); // space between each tab
      const buttonSize = this.$refs.slideRight
        ? this.$refs.slideRight.$el.offsetWidth : this.$refs.slideLeft.offsetWidth;
      const pages = [{ tabIndex: 0, offsetLeft: 0 }]; // beginning scroll position is known
      const tabPages = {};

      for (let i = 0; i < headerElements.length; i += 1) {
        /*
          buttonSpace will double if there are going to be both left and right pagination buttons
        */
        const buttonSpace = pages.length < 2 ? buttonSize : buttonSize * 2;
        const elem = headerElements[i];

        width += elem.offsetWidth; // add elem width but not margin for checks

        if (width > this.headerWidth - buttonSpace) {
          // we have a new page!

          if ((this.headerScrollWidth - totalWidth) < (this.headerWidth - buttonSize)) {
            // there is less that one page remaining!
            pages.push({ tabIndex: i, offsetLeft: endPage });

            // add remaining items to tabPages
            for (let k = i; k < headerElements.length; k += 1) {
              tabPages[k] = pages.length - 1;
            }
            break;
          }

          // more than one page remaining
          pages.push({ tabIndex: i, offsetLeft: elem.offsetLeft - buttonSize }); // align where left scroll button ends
          width = elem.offsetWidth; // reset for new page
        }

        width += tabSpace; // add margin for [i] element
        totalWidth += (elem.offsetWidth + tabSpace);

        tabPages[i] = pages.length - 1;
      }

      this.pages = pages;
      this.tabPages = tabPages;
      this.pageIndex = this.tabPages[this.activeTabIndex];
    },
    updateUnderline() {
      console.log('updateUnderline');
      const elements = Array.from(this.$refs.cdrTabsHeader.children);
      if (elements.length > 0) {
        const activeTab = elements[this.activeTabIndex];
        this.underlineOffsetX = activeTab.offsetLeft + Number(this.leftPosition);
        this.underlineWidth = this.tabPages[this.activeTabIndex] !== this.pageIndex
          ? 0 : activeTab.firstChild.offsetWidth;
      }
    },
    animationEnd(event) {
      if (event.propertyName === 'left') {
        this.updateUnderline();
      }
    },
    handleDownArrowNav() {
      if (!this.animationInProgress) {
        this.$el.lastElementChild.children[this.activeTabIndex].focus();
      }
    },
    setFocusToActiveTabHeader() {
      this.$refs.cdrTabsHeader.children[this.activeTabIndex].children[0].focus();
    },
    getTabEl(tab) {
      return tab.disabled ? (
        <span
          class={clsx(
            this.style['cdr-tabs__header-item-label'],
            this.style['cdr-tabs__header-item-label--disabled'],
          )}
          aria-disabled="true"
          aria-selected="false"
        >
          {tab.name}
        </span>
      ) : (
        <a
          tabIndex={tab.active ? 0 : -1}
          vOn:click_prevent={e => this.handleClick(tab, e)}
          href={`#${tab.id}`}
          class={this.style['cdr-tabs__header-item-label']}
        >
          { tab.name }
        </a>
      );
    },
  },
  render() {
    return (
      <div
        class={clsx(this.style[this.baseClass], this.modifierClass, this.sizeClass)}
        ref="cdrTabsContainer"
        style={{ height: this.height }}
        vOn:transitionend={this.animationEnd}
      >
        <div
          class={clsx(
            this.style['cdr-tabs__gradient-container'],
          )}
          vOn:keyup_right={this.rightArrowNav}
          vOn:keyup_left={this.leftArrowNav}
          vOn:keydown_down_prevent={this.handleDownArrowNav}
        >
          <cdr-button
            icon-only
            with-background={true}
            aria-label=""
            tabIndex="-1"
            vOn:click={this.slideLeft}
            ref="slideLeft"
            class={clsx(
              this.style['cdr-tabs__button'],
              this.style['cdr-tabs__nav-scroll-left'],
              { 'cdr-tabs__button--active': this.overflowLeft },
            )}
          >
            <cdr-icon
              use="#caret-left"
              inherit-color
              slot="icon"
              class="cdr-button__icon"
            />
          </cdr-button>

          <cdr-button
            icon-only
            with-background={true}
            aria-label=""
            tabIndex="-1"
            vOn:click={this.slideRight}
            ref="slideRight"
            class={clsx(
              this.style['cdr-tabs__button'],
              this.style['cdr-tabs__nav-scroll-right'],
              { 'cdr-tabs__button--active': this.overflowRight },
            )}
          >
            <cdr-icon
              use="#caret-right"
              inherit-color
              slot="icon"
              class="cdr-button__icon"
            />
          </cdr-button>

          <nav
            class={clsx(
              this.style['cdr-tabs__header-container'],
            )}
            ref="cdrTabsHeaderContainer"
          >
            <ol
              class={this.style['cdr-tabs__header']}
              role="tablist"
              ref="cdrTabsHeader"
              style={ { left: `${this.leftPosition}px` } }
            >
              {this.tabs.map(tab => (
                  <li
                    role="tab"
                    aria-selected={tab.active}
                    aria-disabled="false"
                    aria-controls={tab.id}
                    id={tab.ariaLabelledby}
                    key={tab.id}
                    class={clsx(
                      tab.active ? this.style['cdr-tabs__header-item-active'] : '',
                      this.style['cdr-tabs__header-item'],
                    )}
                  >
                    {this.getTabEl(tab)}
                  </li>
              ))}
            </ol>
          </nav>

          <div
            class={this.style['cdr-tabs__underline']}
            style={this.underlineStyle}
          />
        </div>
        <div
          class={this.style['cdr-tabs__content-container']}
          ref="slotWrapper"
        >
          {this.$slots.default}
        </div>
      </div>
    );
  },
};
