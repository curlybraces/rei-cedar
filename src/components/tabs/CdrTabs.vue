<template>
  <!-- disable lint errors on line length in template -->
  <!-- eslint-disable max-len -->
  <div :class="[modifierClass]">
    <ol
      :class="$style['cdr-tabs__header']"
      ref="cdrTabsHeader">
      <li
        v-for="tab in tabs"
        :key="tab.id"
        :class="[ tab.active ? $style['cdr-tabs__header-item-active'] : '', $style['cdr-tabs__header-item']]"
        @click="handleClick(tab, $event)">
        <a>
          {{ tab.name }}
        </a>
      </li>
    </ol>
    <hr
      :class="$style['cdr-tabs__underline']"
      :style="underlineStyle">
    <div :class="$style['cdr-tabs__content-container']">
      <slot/>
    </div>
  </div>
</template>

<script>
import modifier from 'mixinsdir/modifier';

export default {
  name: 'CdrTabs',
  mixins: [modifier],
  data() {
    return {
      tabs: [],
      underlineOffsetX: 0,
      underlineOffsetWidth: 0,
      widthInitialized: false,
      activeTabIndex: 0,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-tabs';
    },
    underlineStyle() {
      return {
        marginLeft: `${this.underlineOffsetX}px`,
        width: `${this.underlineOffsetWidth}px`,
      };
    },
  },
  beforeMount() {
    this.tabs = this.$children;
  },
  mounted() {
    if (this.tabs.length > 0) this.tabs[0].setActive(true);
  },
  updated() {
    this.initializeOffsets();
  },
  methods: {
    handleClick(tabClicked, event) {
      const newSelectedTab = this.tabs.find(tab => tabClicked.name === tab.name);
      this.tabs.forEach((tab, index) => {
        if (newSelectedTab.name === tab.name) {
          if (this.activeTabIndex < index) {
            tab.setAnimationDirection('flyRight');
            this.tabs[this.activeTabIndex].setAnimationDirection('flyLeft');
          } else {
            tab.setAnimationDirection('flyLeft');
            this.tabs[this.activeTabIndex].setAnimationDirection('flyRight');
          }
          this.activeTabIndex = index;
          this.$nextTick(tab.setActive(true));
        } else {
          this.$nextTick(tab.setActive(false));
        }
      });
      this.underlineOffsetX =
        event.currentTarget.offsetLeft
        - event.currentTarget.parentElement.parentElement.offsetLeft;
      this.underlineOffsetWidth = event.currentTarget.offsetWidth;
    },
    initializeOffsets() {
      if (!this.widthInitialized) {
        const elements = Array.from(this.$refs.cdrTabsHeader.children);
        this.underlineOffsetWidth = elements[0].offsetWidth;
        this.widthInitialized = true;
      // elements.forEach((element, index) => {
        //   this.tabs[index].setOffsetX(offsetX);
      //   offsetX += element.offsetWidth;
      // });
      }
    },
  },
};
</script>

<style module>
  @import '../../css/settings/_index.pcss';
  @import './styles/CdrTabs.pcss';
</style>
