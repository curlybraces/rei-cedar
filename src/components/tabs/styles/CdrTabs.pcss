/* ==========================================================================
  # CdrTabs

  All values should map to variables in
  vars/CdrTabs.vars.pcss
  in order to allow for theming

  TOC:

    :Base - CdrTabs
    :Elements
      :header
      :header-item
========================================================================== */

.cdr-tabs__header {
  width: 100%;
  display: inline-flex;
  align-content: stretch;
}
