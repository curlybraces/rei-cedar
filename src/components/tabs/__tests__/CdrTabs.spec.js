import { mount } from '@vue/test-utils';
import CdrTabs from 'componentdir/tabs/CdrTabs';
import CdrTabPanel from 'componentdir/tabs/CdrTabPanel';
import Vue from 'vue';

// Tests use nextTick because of the nextTick in mounted hook of tabs

describe('CdrTabs', () => {
  describe('mounted', () => {
    it('mounts tabs', () => {
      const wrapper = mount(CdrTabs);
      expect(wrapper.element).toMatchSnapshot();
    });

    it('mounts with cdr-tab-panel children', (done) => {
      const spyInitActiveTabIndex = jest.fn();
      const spyGetHeaderWidth = jest.fn();
      const spyCalculateOverflow = jest.fn();
      const spyUpdateUnderline = jest.fn();

      const wrapper = mount(CdrTabs, {
        stubs: {
          'cdr-tab-panel': CdrTabPanel,
        },
        slots: {
          default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>']
        },
        methods: {
          initActiveTabIndex: spyInitActiveTabIndex,
          getHeaderWidth: spyGetHeaderWidth,
          calculateOverflow: spyCalculateOverflow,
          updateUnderline: spyUpdateUnderline,
        },
      });

      Vue.nextTick(() => {
        expect(wrapper.vm.tabs.length).toBe(2);
        expect(wrapper.findAll('li').length).toBe(2);
        expect(wrapper.element).toMatchSnapshot();
        expect(spyInitActiveTabIndex).toHaveBeenCalled();
        expect(spyGetHeaderWidth).toHaveBeenCalled();
        expect(spyCalculateOverflow).toHaveBeenCalled();
        
        setTimeout(() => {
          expect(spyUpdateUnderline).toHaveBeenCalled();
          done();
        }, 200)
      });
    });
  });

  describe('event listeners', () => {    
    it('handles scroll event', (done) => {
      const spyCalculateOverflow = jest.fn();
      const spyUpdateUnderline = jest.fn();

      const wrapper = mount(CdrTabs, {
        stubs: {
          'cdr-tab-panel': CdrTabPanel,
        },
        slots: {
          default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>'],
        },
        methods: {
          calculateOverflow: spyCalculateOverflow,
          updateUnderline: spyUpdateUnderline,
        },
        attachToDocument: true,
      });

      Vue.nextTick(() => {
        wrapper.vm.$refs.cdrTabsHeader.parentElement.dispatchEvent(new Event('scroll'));

        setTimeout(() => { // for debounce
          expect(wrapper.vm.overflowLeft).toBe(false);
          expect(spyCalculateOverflow).toHaveBeenCalled();
          expect(spyUpdateUnderline).toHaveBeenCalled();
          wrapper.destroy();
          done();
        }, 600);
      });
    });

    it('handles resize event', () => {
      const spyCalculateOverflow = jest.fn();
      const spyUpdateUnderline = jest.fn();
      const spyGetHeaderWidth = jest.fn();

      const wrapper = mount(CdrTabs, {
        stubs: {
          'cdr-tab-panel': CdrTabPanel,
        },
        slots: {
          default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>'],
        },
        methods: {
          calculateOverflow: spyCalculateOverflow,
          updateUnderline: spyUpdateUnderline,
          getHeaderWidth: spyGetHeaderWidth,
        },
        attachToDocument: true,
      });

      Vue.nextTick(() => {
        setTimeout(() => {
          window.dispatchEvent(new Event('resize'));

          setTimeout(() => { // for debounce
            expect(spyGetHeaderWidth).toHaveBeenCalled();
            expect(spyCalculateOverflow).toHaveBeenCalled();
            expect(spyUpdateUnderline).toHaveBeenCalled();
            wrapper.destroy();
            done();
          }, 600);
        }, 800);
      });
    });
  });

  describe('initActiveTabIndex', () => {
    it('sets activeTabIndex to activateTab', (done) => {
      const wrapper = mount(CdrTabs, {
        stubs: {
          'cdr-tab-panel': CdrTabPanel,
        },
        slots: {
          default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2" />'],
        },
        propsData: {
          activateTab: 1,
        },
      });

      Vue.nextTick(() => {
        expect(wrapper.vm.activeTabIndex).toBe(1);
        done();
      });
    }); 

    it('calls getNextTab', () => {
      const spyGetNextTab = jest.fn();

      const wrapper = mount(CdrTabs, {
        stubs: {
          'cdr-tab-panel': CdrTabPanel,
        },
        slots: {
          default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2" />'],
        },
        methods: {
          getNextTab: spyGetNextTab,
        }
      });

      expect(spyGetNextTab).toHaveBeenCalled();
    });
  });

  it('getNextTab and getPreviousTab', () => {
    const wrapper = mount(CdrTabs, {
      stubs: {
        'cdr-tab-panel': CdrTabPanel,
      },
      slots: {
        default: [
          '<cdr-tab-panel name="tab1"/>',
          '<cdr-tab-panel name="tab2" :disabled="true" />',
          '<cdr-tab-panel name="tab3" />',
        ]
      }
    });

    expect(wrapper.vm.getNextTab(0)).toBe(0);
    expect(wrapper.vm.getNextTab(1)).toBe(2);
    expect(wrapper.vm.getNextTab(3)).toBe(-1);

    expect(wrapper.vm.getPreviousTab(0)).toBe(0);
    expect(wrapper.vm.getPreviousTab(1)).toBe(0);
    expect(wrapper.vm.getPreviousTab(-1)).toBe(-1);
  });

  describe('handleClick', () => {
    it('left to right', (done) => {
      const wrapper = mount(CdrTabs, {
        stubs: {
          'cdr-tab-panel': CdrTabPanel,
        },
        slots: {
          default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>']
        },
      });
      
      Vue.nextTick(() => {
        wrapper.findAll('a').at(1).trigger('click');
        expect(wrapper.vm.activeTabIndex).toBe(1);
        done();
      });
    });

    xit('right to left', (done) => {
      const wrapper = mount(CdrTabs, {
        stubs: {
          'cdr-tab-panel': CdrTabPanel,
        },
        slots: {
          default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>']
        },
        propsData: {
          activateTab: 1,
        },
      });
      
      Vue.nextTick(() => {
        const test = wrapper.findAll('a').at(0);
        console.log('test', test);
        // wrapper.findAll('a').at(0).trigger('click');
        // expect(wrapper.vm.activeTabIndex).toBe(1);
        done();
      });
    });
  });

  it('handles right arrow key', (done) => {
    const wrapper = mount(CdrTabs, {
      stubs: {
        'cdr-tab-panel': CdrTabPanel,
      },
      slots: {
        default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>']
      }
    });
    
    Vue.nextTick(() => {
      // Trigger right arrow keyup event
      wrapper.findAll('div').at(1).trigger('keyup.right');
      expect(wrapper.vm.activeTabIndex).toBe(1);
      done();
    });
  });

  it('handles left arrow key', (done) => {
    const wrapper = mount(CdrTabs, {
      stubs: {
        'cdr-tab-panel': CdrTabPanel,
      },
      slots: {
        default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>']
      }
    });
    
    Vue.nextTick(() => {
      wrapper.setData({ activeTabIndex: 1 });
      // Trigger left arrow keyup event
      wrapper.findAll('div').at(1).trigger('keyup.left');
      expect(wrapper.vm.activeTabIndex).toBe(0);
      done();
    });
  });

  xit('resize calculates overflow properly', (done) => {
    const spyCalculateOverflow = jest.fn();
    const wrapper = mount(CdrTabs, {
      stubs: {
        'cdr-tab-panel': CdrTabPanel,
      },
      slots: {
        default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>']
      },
      methods: {
        calculateOverflow: spyCalculateOverflow,
      },
      attachToDocument: true,
    });
    
    Vue.nextTick(() => {
      window.dispatchEvent(new Event('resize'));
      setTimeout(() => { // for debounce
        expect(wrapper.vm.overflowLeft).toBe(false);
        expect(spy).toHaveBeenCalled();
        wrapper.destroy();
        done();
      }, 600);
    });
  });

  


  // xit('handles right arrow key when far right tab is active', (done) => {
  //   const wrapper = mount(CdrTabs, {
  //     stubs: {
  //       'cdr-tab-panel': CdrTabPanel,
  //     },
  //     slots: {
  //       default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>']
  //     }
  //   });
  //   Vue.config.errorHandler = done
  //   Vue.nextTick(() => {
  //     wrapper.setData({ activeTabIndex: 1 });
  //     // Trigger right arrow keyup event
  //     wrapper.findAll('div').at(1).trigger('keyup.right');
  //     expect(wrapper.vm.activeTabIndex).toBe(1);
  //     done();
  //   });
  // });



  // xit('handles left arrow key when far left is active', (done) => {
  //   const wrapper = mount(CdrTabs, {
  //     stubs: {
  //       'cdr-tab-panel': CdrTabPanel,
  //     },
  //     slots: {
  //       default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>']
  //     }
  //   });
  //   Vue.config.errorHandler = done
  //   Vue.nextTick(() => {
  //     // Trigger left arrow keyup event
  //     wrapper.findAll('div').at(1).trigger('keyup.left');
  //     expect(wrapper.vm.activeTabIndex).toBe(0);
  //     done();
  //   });
  // });

  // xit('ignores non left and right arrow key', (done) => {
  //   const wrapper = mount(CdrTabs, {
  //     stubs: {
  //       'cdr-tab-panel': CdrTabPanel,
  //     },
  //     slots: {
  //       default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>']
  //     }
  //   });
  //   Vue.config.errorHandler = done
  //   Vue.nextTick(() => {
  //     // Trigger down arrow keyup event
  //     wrapper.findAll('div').at(1).trigger('keyup.down');
  //     expect(wrapper.vm.activeTabIndex).toBe(0);
  //     done();
  //   });
  // });




  // it('click tab changes active tab lower index variation', (done) => {
  //   const wrapper = mount(CdrTabs, {
  //     stubs: {
  //       'cdr-tab-panel': CdrTabPanel,
  //     },
  //     slots: {
  //       default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>']
  //     },
  //   });
  //   Vue.config.errorHandler = done;

  //   wrapper.setData({ activeTabIndex: 1 });

  //   Vue.nextTick(() => {
  //     setTimeout(() => {
  //       wrapper.findAll('a').at(0).trigger('click');
  //       expect(wrapper.vm.activeTabIndex).toBe(0);

  //       wrapper.findAll('a').at(1).trigger('click');
  //       expect(wrapper.vm.activeTabIndex).toBe(1);
  //     }, 500);

  //     done()
  //   })
  // });

  // xit('calculateOverflow sets header overflow properly', () => {
  //   const wrapper = mount(CdrTabs, {
  //     stubs: {
  //       'cdr-tab-panel': CdrTabPanel,
  //     },
  //     slots: {
  //       default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>']
  //     }
  //   });
  //   wrapper.setData({ headerWidth: 2000 });
  //   wrapper.vm.calculateOverflow();
  //   expect(wrapper.vm.headerOverflow).toBe(true);
  // });

  // xit('scrollbar is hidden properly', (done) => {
  //   const wrapper = mount(CdrTabs, {
  //     stubs: {
  //       'cdr-tab-panel': CdrTabPanel,
  //     },
  //     slots: {
  //       default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>']
  //     },
  //     attachToDocument: true,
  //   });
  //   Vue.config.errorHandler = done;
  //   wrapper.setData({ widthInitialized: true});
  //   wrapper.setData({ underlineWidth: -1});
  //   wrapper.vm.hideScrollBar();
  //   expect(wrapper.vm.$refs.cdrTabsContainer.style.getPropertyValue('overflow-x')).toBe('hidden');
  //   window.dispatchEvent(new Event('transitionend'));
  //   Vue.nextTick(() => {
  //     expect(wrapper.vm.$refs.cdrTabsContainer.style.getPropertyValue('overflow-x')).toBe('unset');
  //     wrapper.destroy();
  //     done();
  //   });
  // });

  // xit('handles down arrow', (done) => {
  //   const wrapper = mount(CdrTabs, {
  //     stubs: {
  //       'cdr-tab-panel': CdrTabPanel,
  //     },
  //     slots: {
  //       default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>']
  //     }
  //   });
  //   Vue.config.errorHandler = done;

  //   const spy = spyOn(wrapper.vm, 'handleDownArrowNav')
  //   Vue.nextTick(() => {
  //     // Trigger right arrow keyup event
  //     wrapper.findAll('div').at(1).trigger('keydown.down');
  //     expect(spy).toHaveBeenCalled();
  //     done();
  //   });
  // });

  // xit('handles up arrow', (done) => {
  //   const wrapper = mount(CdrTabs, {
  //     stubs: {
  //       'cdr-tab-panel': CdrTabPanel,
  //     },
  //     slots: {
  //       default: ['<cdr-tab-panel name="tab1"/>', '<cdr-tab-panel name="tab2"/>']
  //     }
  //   });
  //   Vue.config.errorHandler = done;

  //   const spy = spyOn(wrapper.vm, 'setFocusToActiveTabHeader');
  //   Vue.nextTick(() => {
  //     wrapper.find({ ref: 'slotWrapper' }).findAll('div').at(1).trigger('keydown.up');
  //     expect(spy).toHaveBeenCalled();
  //     done();
  //   });
  // });

});
