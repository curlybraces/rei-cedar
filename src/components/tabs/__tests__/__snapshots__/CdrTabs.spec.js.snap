// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrTabs mounted mounts tabs 1`] = `
<div
  class="cdr-tabs"
  style="height: 240px;"
>
  <div
    class="cdr-tabs__gradient-container"
  >
    <nav
      class="cdr-tabs__header-container"
    >
      <ol
        class="cdr-tabs__header"
        role="tablist"
      />
    </nav>
    <div
      class="cdr-tabs__underline"
      style="transform: translateX(0px); width: 0px;"
    />
  </div>
  <div
    class="cdr-tabs__content-container"
  />
</div>
`;

exports[`CdrTabs mounted mounts with cdr-tab-panel children 1`] = `
<div
  class="cdr-tabs"
  style="height: 240px;"
>
  <div
    class="cdr-tabs__gradient-container"
  >
    <nav
      class="cdr-tabs__header-container"
    >
      <ol
        class="cdr-tabs__header"
        role="tablist"
      >
        <li
          class="cdr-tabs__header-item"
        >
          <a
            aria-controls="tab1"
            aria-disabled="false"
            class="cdr-tabs__header-item-label"
            href="#tab1"
            role="tab"
            tabindex="-1"
          >
            tab1
          </a>
        </li>
        <li
          class="cdr-tabs__header-item"
        >
          <a
            aria-controls="tab2"
            aria-disabled="false"
            class="cdr-tabs__header-item-label"
            href="#tab2"
            role="tab"
            tabindex="-1"
          >
            tab2
          </a>
        </li>
      </ol>
    </nav>
    <div
      class="cdr-tabs__underline"
      style="transform: translateX(0px); width: 0px;"
    />
  </div>
  <div
    class="cdr-tabs__content-container"
  >
    <div
      aria-hidden="true"
      class="cdr-tab-panel"
      hidden="hidden"
      id="tab1"
      role="tabpanel"
      tabindex="0"
    />
    <div
      aria-hidden="true"
      class="cdr-tab-panel"
      hidden="hidden"
      id="tab2"
      role="tabpanel"
      tabindex="0"
    />
  </div>
</div>
`;
